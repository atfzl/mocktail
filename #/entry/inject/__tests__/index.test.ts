import 'isomorphic-fetch';
import '..';

import nanoid = require('nanoid');

const env = process.env;

let postMessage: jest.Mock<any, any>;

// mock for nanoid to return same if for all
jest.mock('nanoid');
(nanoid as any).mockImplementation(() => 1);

describe('Inject Script', () => {
  const makeXHR = (...args: Parameters<XMLHttpRequest['open']>) => {
    const xhr = new XMLHttpRequest();
    return new Promise((resolve, reject) => {
      xhr.open(...args);
      xhr.send();
      xhr.onload = resolve;
      xhr.onerror = reject;
    });
  };

  beforeEach(() => {
    postMessage = (global as any).postMessage = jest.fn();
  });

  test('XMLHTTPRequests: postMessage called twice', done => {
    makeXHR('GET', env.apiUrl + '/json', true)
      .then(() => {
        expect(postMessage).toBeCalledTimes(2);
        done();
      })
      .catch(console.error);
  });

  test('Fetch: postMessage called twice', done => {
    fetch(env.apiUrl + '/json')
      .then(() => {
        expect(postMessage).toBeCalledTimes(2);
        done();
      })
      .catch(console.error);
  });

  test('XMLHTTPRequests: Hook snapshot', done => {
    makeXHR('GET', env.apiUrl + '/json', true)
      .then(() => {
        expect(postMessage.mock.calls[0]).toMatchSnapshot();
        expect(postMessage.mock.calls[1]).toMatchSnapshot();
        done();
      })
      .catch(console.error);
  });

  test('Fetch: Hook snapshot', done => {
    fetch(env.apiUrl + '/json', { cache: 'no-cache' })
      .then(() => {
        expect(postMessage.mock.calls[0]).toMatchSnapshot();
        expect(postMessage.mock.calls[1]).toMatchSnapshot();
        done();
      })
      .catch(console.error);
  });
});
