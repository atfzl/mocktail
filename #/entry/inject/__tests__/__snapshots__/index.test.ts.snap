// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inject Script Fetch: Hook snapshot 1`] = `
Array [
  Object {
    "id": 1,
    "payload": Object {
      "body": Object {
        "body": undefined,
        "headers": Object {},
        "method": "GET",
        "url": "http://localhost:5555/json",
      },
      "evt": "before",
    },
    "source": "xhook-atfzl-message",
    "type": "request",
  },
  "*",
]
`;

exports[`Inject Script Fetch: Hook snapshot 2`] = `
Array [
  Object {
    "id": 1,
    "payload": Object {
      "body": Object {
        "request": Object {
          "body": undefined,
          "headers": Object {},
          "method": "GET",
          "url": "http://localhost:5555/json",
        },
        "response": Object {
          "data": Object {
            "name": "john",
          },
          "finalUrl": "http://localhost:5555/json",
          "headers": Object {
            "access-control-allow-headers": "Origin, X-Requested-With, Content-Type, Accept",
            "access-control-allow-origin": "*",
            "connection": "close",
            "content-length": "15",
            "content-type": "application/json; charset=utf-8",
            "date": "1",
            "etag": "1",
            "x-powered-by": "Express",
          },
          "status": 200,
        },
      },
      "evt": "after",
    },
    "source": "xhook-atfzl-message",
    "type": "request",
  },
  "*",
]
`;

exports[`Inject Script XMLHTTPRequests: Hook snapshot 1`] = `
Array [
  Object {
    "id": 1,
    "payload": Object {
      "body": Object {
        "body": undefined,
        "headers": Object {},
        "method": "GET",
        "url": "http://localhost:5555/json",
      },
      "evt": "before",
    },
    "source": "xhook-atfzl-message",
    "type": "request",
  },
  "*",
]
`;

exports[`Inject Script XMLHTTPRequests: Hook snapshot 2`] = `
Array [
  Object {
    "id": 1,
    "payload": Object {
      "body": Object {
        "request": Object {
          "body": undefined,
          "headers": Object {},
          "method": "GET",
          "url": "http://localhost:5555/json",
        },
        "response": Object {
          "data": "{\\"name\\":\\"john\\"}",
          "finalUrl": "http://localhost:5555/json",
          "headers": Object {
            "content-type": "application/json; charset=utf-8",
          },
          "status": 200,
        },
      },
      "evt": "after",
    },
    "source": "xhook-atfzl-message",
    "type": "request",
  },
  "*",
]
`;
